clients/da-vinci-client/src/main/java/com/linkedin/davinci/DaVinciBackend.java

```diff
@@ -576,8 +578,35 @@ protected void reportPushStatus(
     VersionBackend versionBackend = versionByTopicMap.get(kafkaTopic);
     if (versionBackend != null && versionBackend.isReportingPushStatus()) {
       Version version = versionBackend.getVersion();
-      pushStatusStoreWriter
-          .writePushStatus(version.getStoreName(), version.getNumber(), partition, status, incrementalPushVersion);
+      if (writeBatchingPushStatus && !incrementalPushVersion.isPresent()) {
+        // Batching the push statuses from all partitions for batch pushes
+        versionBackend.updatePartitionStatus(partition, status);
+        if (status == ExecutionStatus.COMPLETED) {
+          if (!versionBackend.isBatchPushEndSignalSent()
+              && versionBackend.areAllPartitionsOnSameTerminalStatus(status)) {
+            pushStatusStoreWriter.writeVersionLevelPushStatus(
+                version.getStoreName(),
+                version.getNumber(),
+                status,
+                versionBackend.getTrackedPartitions());
+            versionBackend.batchPushEndSignalSent();
```

Comment:
Consider the scenario where 
1. Get a signal for partition 1 started
2. Get a signal for partition 1 completed
3. Get a signal for partition 2 started
4. Get a signal for partition 2 completed

The above can result in partially sending the completed status w/ only partition 1 information but none for partition 2. 

Is it the above possible? Is there a way to know the all the partitions ahead in time during the `versionBackend` creation?

Reply:
Sync up offline; the race condition has been discussed before, users could subscribe to a new partition anytime; I will try to add a 5 minutes delay before sending the COMPLETE event.

It will only work for batch-push by using the future version subscription callback to send push status event; the current approach can be extended to support incremental push status updates in future. We will keep the current implementation.